### ISSUE TITLE
Operator uses template default `LeaderElectionID`

### ISSUE THREAD

**Issue by stevehipwell** [tigera/operator#1594]
## Expected Behavior

Calling `kubectl -n tigera-operator get lease` should return a lease named for the operator and not a generically named one.

```shell
NAME            HOLDER                                                                  AGE
tigera-operator-lock   tigera-operator-5dcfb9df8c-bfkvv_31aa92d3-9894-4720-88b0-9414067296b3   89d
```

## Current Behavior

The [LeaderElectionID](https://github.com/tigera/operator/blob/master/main.go#L201) value hasn't been customized so we get a generic lock name that could collide with another operator and is ambiguous.

```shell
NAME            HOLDER                                                                  AGE
operator-lock   tigera-operator-5dcfb9df8c-bfkvv_31aa92d3-9894-4720-88b0-9414067296b3   89d
```

## Possible Solution

I'd be happy to open a PR to rename the `LeaderElectionID` value to `tigera-operator-lock`.

## Steps to Reproduce (for bugs)

n/a

## Context

n/a

## Your Environment

n/a

**Comment by tmjd** [comment]
Since the lease is namespaced, in `tigera-operator`, there is not really much chance to conflict with another operator. If it did my first question would be why is it necessary for another operator to be running in the tigera-operator namespace.
Not needing to specify unique resource names is one of the benefits of using namespaces.

If you have a use case of why we would need the lock to be unique please provide that and we can re-evaluate with that info.

**Comment by stevehipwell** [comment]
@tmjd I get that if you only ever install the operator into the `tigera-operator` namespace it's not a major pain point, but it is still limiting for automation as at least two columns would need to be carried to identify it.

However as it's possible to install the operator in an arbitrary namespace at which point this is more of a pain point. I know that in the standard deploy the name isn't high value, but it wouldn't hurt to have a more descriptive name for this use case and for more complex deployments it's worth changing it.

**Comment by tmjd** [comment]
>it is still limiting for automation as at least two columns would need to be carried to identify it.

Why does automation need to identify it? It is an internal resource used only by the operator. (I also think namespaced resources should be differentiated by their namespaces.) If the operator was more of a 'normal' component that could have multiple instances in a cluster and had a lease, I would expect that the leases would only be differentiated by namespace, so I'm not sure why operator should be trying to make its lease name unique when it is unique in its namespace.

>it's possible to install the operator in an arbitrary namespace at which point this is more of a pain point

The operator might be able to be installed in other namespaces but it should be in its own (dedicated) namespace, if it is sharing the namespace with other components/operators then that would be an unsupported install and against recommendation. We can document that the operator should be ran its own namespace if this is not clear but it seems like a reasonable expectation.

The operator is an opinionated install and as such I don't think we are interested in ensuring it can be ran in a namespace with other operators/components. In my opinion I feel that including `tigera-` in resource names in the operator namespace is redundant since they are in the namespace being used for the operator (since the default is tigera-operator).
@caseydavenport WDYT?

**Comment by stevehipwell** [comment]
@tmjd adding a `tigera-` prefix to the lease name wouldn't have any impact to any of your points above as it'd still be a fixed name, just clearer in purpose.

For automation when I run `kubectl get lease -A` it's be much clearer if I had unique lease names, and as I'm suggesting it still be fixed if I get `tigera-operator-lease` I know exactly what it is.

What is the reasoning behind only supporting running the operator in a fixed namespace? We've already proved that it's possible, and in some cases desirable, to run it in a custom namespace with no side effects (in my Helm chart the chart can only be installed once no matter if the namespace is changed).

Finally, what would be the harm in renaming the lock to help with this? It makes zero material difference to the suggested operation mode and it seems to be pretty standard for this to be renamed, even in operators that can only be installed once.

**Comment by tmjd** [comment]
>What is the reasoning behind only supporting running the operator in a fixed namespace?

I don't think I'm suggesting that, I'm suggesting the operator should be in a namespace by itself. Not along with other components or operators, hence no chance of the lock/lease conflicting with anything else.

The immediate harm renaming the lock would do, on upgrade the lock would not be honored or we would need additional code to understand both the `operator-lock` and `tigera-operator-lock`. Also there would be similar downgrade issues that could not be resolved in the operator.
I don't see options with the controller-runtime to support multiple possible names for the lock, so we'd probably have to include some custom logic to ensure upgrades honor the lock appropriately.

**Comment by stevehipwell** [comment]
@tmjd let me rephrase my question, why would the operator need to be in an namespace on it's own? The only reason I can see is the lease name, which would only break if another operator without a customised lease name was also being installed in the same namespace.

On further reflection I can see the issues with a rename at this point. To update the lease name you'd either need to require scaling the current deployment down to 0 via a single pre-install command or changing the code to work with both leases; and neither of these are great options. The least worse option would probably be an optional command line argument to change the name of the lease, which would be a nice to have. I suspect that you're not interested in a PR for this?

**Comment by caseydavenport** [comment]
Yeah, this is the entire purpose of namespaces - to allow re-use of names within a cluster without conflicting. The design for the calico operator is that it is namespaced - this is generally a good practice in terms of permission boundaries and for other operational reasons as well.

I don't currently think we should make this change to the lock name as it breaks upgrade / downgrade compatibility, and I have yet to see a real justification for running additional applications within the calico operator's namespace. If there's a real use-case that can't be solved by just deploying things in a separate namespace, that might provide justification. But, I would say that the more complex the cluster, the more important it is to use namespaces properly and limit the scope of what is deployed in each.

> The least worse option would probably be an optional command line argument to change the name of the lease, which would be a nice to have. I suspect that you're not interested in a PR for this?

Yeah, this is the most I can imagine us doing, but this still feels like a way folks can shoot themselves in the foot (e.g., setting this flag, then forgetting to update it when applying a new operator.yaml and ending up with two operator locks), so I'm not super keen on it.

**Comment by stevehipwell** [comment]
@caseydavenport I don't disagree with the use of namespaces for isolation, but I do disagree with using un-informative names especially ones straight from a template; taking this pattern to the nth degree we'd have the deployment called something like `default`. I think the point here is that the Tigera Operator doesn't need to be in a fully isolated namespace and that namespace definitely doesn't need to be called `tigera-operator` so the current lease name is at best ambiguous.

All of that said, I can't see a way to change this now that the operator is in use without either significant effort or a breaking change; so I'm going to close this issue.

I would like to add a final comment that if there is ever a v2 operator it might be worth looking into using a lease in `kube-system` (I think Cilium do this) as this would guarantee only a single operator could ever be active. This combined with a validating installation method, such as Helm or SSA with field ownership, would significantly improve your single installation stance at the same time.
